### PR TITLE
DAOS-2906 tests: Using ClusterShell in agent and server utils

### PR DESCRIPTION
**Pull request by phender**
Making use of ClusterShell and clush to parallize agent and server
setup and tear down tasks.  When starting multiple agents or servers
we should see decrease in functional test runtime.

Test-tag: regression,vm pooldestroy,quick

Signed-off-by: Phil Henderson <phillip.henderson@intel.com>

### CODE REVIEW

File: src/tests/ftest/util/agent_utils.py [hunk 1]
```diff
@@ -144,59 +110,59 @@ def run_agent(basepath, server_list, client_list=None):
...
         sessions[client].wait()
 
-    # check to make sure it's dead
+    # Check to make sure all the daos agents are dead
+    # pgrep exit status:
+    #   0 - One or more processes matched the criteria.
+    #   1 - No processes matched.
+    #   2 - Syntax error in the command line.
+    #   3 - Fatal error: out of memory etc.
     time.sleep(5)
```
Comment (brianjmurrell) [inline src/tests/ftest/util/agent_utils.py]: Not a fan of this kind of arbitrary sleep (and let's hope we slept long enough).  Can we replace this with a query/wait loop of what we are waiting for?  Keep in mind that looping locally means lots of expensive `ssh`es vs. looping remotely is done in a single ssh.

File: src/tests/ftest/util/agent_utils.py [hunk 2]
```diff
@@ -144,59 +110,59 @@ def run_agent(basepath, server_list, client_list=None):
...
-    This should kill ssh and the agent.  This is temporary; presuming the
-    agent will deamonize at somepoint and can be started killed more
-    appropriately.
+    """Kill ssh and the agent.
 
-    client_list -- kill daos_agent on these hosts
-    sessions    -- set of subprocess sessions returned by run_agent()
-    """
+    This is temporary; presuming the agent will deamonize at somepoint and can
+    be started killed more appropriately.
```
Comment (brianjmurrell) [inline src/tests/ftest/util/agent_utils.py]: Typo/grammar here, only worth fixing if you have to push again for any other more significant reason.
```suggestion
    be started and killed more appropriately.
```

File: src/tests/ftest/util/agent_utils.py [hunk 3]
```diff
@@ -144,59 +110,59 @@ def run_agent(basepath, server_list, client_list=None):
...
-
-    if found_hosts:
-        raise AgentFailed("Attempt to kill daos_agent processes {} on hosts "
-                          "{} was unsuccessful"
-                          .format(', '.join(stdout.splitlines()), found_hosts))
+    result = clush(client_list, "pgrep 'daos_agent'", None, False)
+    if result.exit_status != 1:
+        raise AgentFailed(
+            "DAOS agent processes detected after attempted stop:\n{}".format(
+                result.stdout))
```
Comment (brianjmurrell) [inline src/tests/ftest/util/agent_utils.py]: Can we easily report which hosts they were on, to help somebody debugging with which logs to look at?

File: src/tests/ftest/util/general_utils.py [hunk 4]
```diff
@@ -44,6 +44,60 @@ class DaosTestError(Exception):
     """DAOS API exception class."""
 
 
+def clush(hostlist, command, env=None, verbose=True):
```
Comment (brianjmurrell) [inline src/tests/ftest/util/general_utils.py]: I wonder if we should make this name more generic so that if we ever replace `clush` in the future with something else, we can just drop that in here and not feel like we need to go updating the name everywhere.

File: src/tests/ftest/util/general_utils.py [hunk 5]
```diff
@@ -44,6 +44,60 @@ class DaosTestError(Exception):
...
+            .duration       duration of the command
+            .interrupted    was the command interrupted
+            .pid            pid of the command
+
+    """
+    nodeset = NodeSet.fromlist(hostlist)
+    clush_command = "clush -w {} -B -S \"{}\"".format(nodeset, command)
+    return process.run(
+        clush_command, env=env, ignore_status=True, shell=True,
+        verbose=verbose, allow_output_check="none")
```
Comment (brianjmurrell) [inline src/tests/ftest/util/general_utils.py]: Given that _ClusterShell_ is primarily a python library, shouldn't we avoid the trip through a `fork()`/`exec()` and the shell here and work with _ClusterShell_ directly like you did for `check_file_exists()`?

File: src/tests/ftest/util/general_utils.py [hunk 6]
```diff
@@ -44,6 +44,60 @@ class DaosTestError(Exception):
...
+        (bool, NodeSet): A tuple of:
+            - True if the file exists on each of the hosts; False otherwise
+            - A NodeSet of hosts on which the file does not exist
+
+    """
+    missing_file = NodeSet()
+    nodeset = NodeSet.fromlist(hosts)
+
+    task = task_self()
+    task.run("test -e {}".format(filename), nodes=nodeset)
```
Comment (brianjmurrell) [inline src/tests/ftest/util/general_utils.py]: Should probably quote the argument to `test -e` here to avoid problems with spaces (and other characters) in the filename.

File: src/tests/ftest/util/server_utils.py [hunk 7]
```diff
@@ -245,49 +290,46 @@ def stop_server(setname=None, hosts=None):
...
 
-    # make sure they actually stopped
-    # but give them some time to stop first
+    # Make sure the servers actually stopped.  Give them time to stop first
+    # pgrep exit status:
+    #   0 - One or more processes matched the criteria.
+    #   1 - No processes matched.
+    #   2 - Syntax error in the command line.
+    #   3 - Fatal error: out of memory etc.
     time.sleep(5)
```
Comment (brianjmurrell) [inline src/tests/ftest/util/server_utils.py]: Ditto for the `sleep()` here.

File: src/tests/ftest/util/general_utils.py (hunk 6 above)
Comment (phender) [reply src/tests/ftest/util/general_utils.py]: Agreed, good catch.

File: src/tests/ftest/util/agent_utils.py (hunk 2 above)
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: Fixed - Thanks.

File: src/tests/ftest/util/agent_utils.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: I completely agree about arbitrary sleeps.  It would be great to know why it was added in the first place.  Given that the `Popen.poll()` checks to see if the child process (daos_agent) has terminated and the `Popen.wait()` waits for the child process to terminate (which should probably be indented with the `Popen.kill()`), I don't see an obvious need for the sleep.  I assume at some point there was a need and removing it could cause potential CI issues even if they are not seen during this PR.  Should the removal of sleeps be a separate PR?

File: src/tests/ftest/util/server_utils.py (hunk 7 above)
Comment (phender) [reply src/tests/ftest/util/server_utils.py]: Again I agree - never been a fan of sleeps (there is another on line 284).  The Popen.poll() and Popen.wait() calls above should handle ensuring the stop commands have completed.  I'm just weary of removing an undocumented piece of code that may handle a corner case that prevents CI failures - especially when it is slightly out of the scope of the PR.

File: src/tests/ftest/util/agent_utils.py (hunk 3 above)
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: Yes.  The new method returns a dictionary of return codes and the nodes with those return codes.

File: src/tests/ftest/util/general_utils.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/util/general_utils.py]: Good suggestion.  Let me know what you think of the new name - I'm up for other suggestions if they make sense.

File: src/tests/ftest/util/general_utils.py (hunk 5 above)
Comment (phender) [reply src/tests/ftest/util/general_utils.py]: Yes.  I've updated the method and in doing so discovered that using the maximum return code is problematic with commands like pkill and pgrep.  The new method returns a dictionary of each return code detected - paired with the nodes that returned the code - to address this issue.

File: src/tests/ftest/util/server_utils.py [hunk 8]
```diff
@@ -245,49 +292,49 @@ def stop_server(setname=None, hosts=None):
...
-    for host in hosts:
-        subprocess.call("ssh {0} \"{1}\"".format(host, '; '.join(kill_cmds)),
-                        shell=True)
+    kill_cmds = [
+        "pkill '(daos_server|daos_io_server)' --signal INT",
+        "sleep 5",
+        "pkill '(daos_server|daos_io_server)' --signal KILL",
+    ]
+    # Intentionally ignoring the exit status of the command
+    cluster_cmd(hosts, "; ".join(kill_cmds), False)
```
Comment (phender) [inline src/tests/ftest/util/server_utils.py]: To avoid this message from being printed:
`boro-[41,43,52-53]: failure running 'pkill '(daos_server|daos_io_server)' --signal INT; sleep 5; pkill '(daos_server|daos_io_server)' --signal KILL': rc=1
`
```suggestion
    cluster_cmd(hosts, "; ".join(kill_cmds), False, None, None)
```

File: src/tests/ftest/util/general_utils.py (hunk 4 above)
Comment (brianjmurrell) [reply src/tests/ftest/util/general_utils.py]: How about `pcmd`?

File: src/tests/ftest/util/agent_utils.py (hunk 1 above)
Comment (brianjmurrell) [reply src/tests/ftest/util/agent_utils.py]: Frankly I suspect it was added "just in case" and that we should remove the for all of the reasons you stated.

I suppose that could be a different PR though.

File: src/tests/ftest/util/general_utils.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/util/general_utils.py]: Done

File: src/tests/ftest/util/agent_utils.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: I think this will be handled by the changes going on in #911

File: src/tests/ftest/util/agent_utils.py [hunk 9]
```diff
@@ -33,91 +33,64 @@
...
 
-    retcode, node, agent_dir = node_setup_okay(server_list, NodeListType.SERVER)
-    if not retcode:
-        raise AgentFailed("Server node " + node + " does not have directory "
-                          + agent_dir + " set up correctly for user "
-                          + user + ".")
-
     # if empty client list, 'self' is effectively client
     if client_list is None:
         client_list = [socket.gethostname().split('.', 1)[0]]
```
Comment (mjean308) [inline src/tests/ftest/util/agent_utils.py]: This assumption is an issue on A21 cluster not in our env;  but it brings up a question on portability.    Why do we assume that that the testnode/VM has daos installed?

Comment (brianjmurrell) [reply src/tests/ftest/util/agent_utils.py]: Because it's the job of whatever runs the tests to ensure that whatever is being tested is already installed.  The tests' job is to run tests, not prepare nodes for the test.

This is why in our CI pipeline, we have processes that install whatever is going to be tested before we call the test running process.

Comment (brianjmurrell) [reply src/tests/ftest/util/agent_utils.py]: Hrm.  In rereading your question, is your question actually "why is the testnode/VM being used as a daos client at all and not just being the test runner only?"

Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: It feels like what is being said here is all tests should explicitly specify at least one client and one server and therefore if line 70 should be reporting an error.  Tests will then be more portable.  This sort of change would benefit from #746 (due to the `-tc` laumch.py option) and should either be part of that PR, or better yet, a new one.

File: src/tests/ftest/util/agent_utils.py (hunk 1 above)
Comment (brianjmurrell) [reply src/tests/ftest/util/agent_utils.py]: I don't see that patch even touching `src/tests/ftest/util/agent_utils.py`.

At the same time, still not saying that this needs to be resolved in this PR.

File: src/tests/ftest/util/agent_utils.py (hunk 9 above)
Comment (brianjmurrell) [reply src/tests/ftest/util/agent_utils.py]: So, yes.  If what @mjean308 is asking is why is the test runner also a daos client (for many tests), I would generally agree that it's best to isolate the code being tested onto a node other than the node that is running the tests.

I'm not sure that's as easy as just defining a client in all tests' YAML files thoufgh.  Many tests have been written to use the Python DAOS API directly so those tests would need to be converted to a style (perhaps some kind of RPC mechanism) where that Python DAOS API client is on a remote node and being remotely controlled by the Avocado test.

There is some support in Avocado for running tests on remote nodes.  I have not investigated exactly what that is but perhaps that could be of benefit here too.

Additionally, there is a slight bit of economics in running the client on the test-runner as it means that all other available (8 in the case of our CI clusters) nodes can be DAOS servers.  Although that could easily be ameliorated by just adding a VM or two more to the hardware clusters to be clients as all of the benefits of hardware nodes relate to servers I think.

Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: > I'm not sure that's as easy as just defining a client in all tests' YAML files though.

Yes, I agree and I did not mean to imply it would be simple.  In my mind that's probably the strongest argument for a unique PR.

> Additionally, there is a slight bit of economics in running the client on the test-runner

Agreed.  The additional VMs would probably be the best option, but nothing would stop the yaml file from specifying the launch node as the client.

File: src/tests/ftest/util/agent_utils.py [hunk 10]
```diff
@@ -33,91 +33,64 @@
...
+    file_checks = (
+        ("Server", server_list, "/var/run/daos_server"),
+        ("Client", client_list, "/var/run/daos_agent"),
+    )
+    for host_type, host_list, directory in file_checks:
+        status, nodeset = check_file_exists(host_list, directory)
+        if not status:
+            raise AgentFailed(
+                "{}: {} missing directory {} for user {}.".format(
+                    nodeset, host_type, directory, user))
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: Carry over from the original code - and no, it does not.  The check should be more robust and check the file permissions using the user.  It may not gain much in CI but could be useful in manual testing for those not bitten once.

File: src/tests/ftest/util/agent_utils.py [hunk 11]
```diff
@@ -140,17 +113,17 @@ def run_agent(basepath, server_list, client_list=None):
...
                 time.sleep(1)
                 continue
             expected_data += output
 
             match = re.findall(pattern, output)
-            if match:
-                started_clients.append(client)
-                print("<AGENT> agent started on node {} in {} "
-                      "seconds".format(client, time.time() - start_time))
+            if len(match) > 0:
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: Interesting.  I would have thought explicit vs implicit would have won out here.  I will fix it.

File: src/tests/ftest/util/agent_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/agent_utils.py]: Added verifying the user ownership as well as the path existence.
